from_json/date_test.ts

```diff
@@ -1,72 +1,81 @@
+  fn() {
+    const testJson = `{"date":"2020-12-31T12:00:00.300Z"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
     assert(testObj.date instanceof Date);
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.300Z");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-12-31T23:00:00+01:00",
+  name: "iso8601Date parses without milliseconds - 2020-12-31T12:00:00Z",
   fn() {
-    const testJson =
-      `{"date":"2020-12-31T23:00:00+01:00","not_a_date":"Hello world"}`;
-    const testObj = JSON.parse(testJson, (_, v) => ISODateFromJson(v));
+    const testJson = `{"date":"2020-12-31T12:00:00"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
     assert(testObj.date instanceof Date);
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z");
   },
 });
 
 test({
-  name: "ISODateFromJson parses ISO dates - 2020-12-31T23:00:00",
+  name: "iso8601Date parses with added timezone - 2020-12-31T12:00:00Z-6",
   fn() {
-    const testJson =
-      `{"date":"2020-12-31T23:00:00","not_a_date":"Hello world"}`;
-    const testObj = JSON.parse(testJson, (_, v) => ISODateFromJson(v));
+    const testJson = `{"date":"2020-12-31T06:00:00Z-6"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
     assert(testObj.date instanceof Date);
-    assertEquals(testObj.date.getFullYear(), 2020);
-    assert(!(testObj.not_a_date instanceof Date));
-    assertEquals(testObj.not_a_date, "Hello world");
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z");
   },
 });
 
 test({
-  name: "Will not serialize non date strings",
+  name: "iso8601Date parses time without `T` - 2020-06-29 12:00:00Z",
   fn() {
-    class Test extends Serializable {
-      @SerializeProperty({
-        fromJsonStrategy: ISODateFromJson,
-      })
-      date!: Date | string;
-    }
-    const testObj = new Test().fromJson(`{"date":"I am not a date!"}`);
-    assertEquals(typeof testObj.date, "string");
-    assertEquals(testObj.date, "I am not a date!");
+    const testJson = `{"date":"2020-12-31 12:00:00Z"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
+    assert(testObj.date instanceof Date);
+    assertEquals(testObj.date.toISOString(), "2020-12-31T12:00:00.000Z");
+  },
+});
+
+test({
+  name: "iso8601Date parses `2020-366` - leap year",
+  fn() {
+    const testJson = `{"date":"2020-366"}`;
+    const testObj = JSON.parse(testJson, (_, v) => iso8601Date(v));
+    assert(testObj.date instanceof Date);
+    assertEquals(testObj.date.toISOString(), "2020-12-31T00:00:00.000Z");
   },
```

Comment:
Can I suggest a test to check that YYYY-MM-DD is deserialized as midnight local time? (I think this test will pass, looking at the code)

This is how we'll want it for our date types. (Code that uses isoDateStringToLocalDate() now)

Here is how we tested it in date-utils.spec.ts:
```
    // 1990-12-11
    const localDateMidnight = new Date(1990, 11, 11);
...
expect(isoDateStringToLocalDate("1990-12-11").getTime()).toEqual(localDateMidnight.getTime());
```

Reply:
This depends on the string, "1990-12-11" will have zones applied to it,  where "1990/12/11". I am going to divert his today ist up to authors to make the strategies they want now.